File: rust/src/pricing/egress_pricing.rs
//! Egress pricing engine — resolves a per-GB egress rate from
//! `(provider, region)` using the bundled `data/egress_prices.json` catalog.
//!
//! Mirrors the Python [`dexcost.egress_pricing`] module shape: bundled JSON +
//! a resolver returning `(rate, pricing_source, cost_confidence)`.
//!
//! Fail-silent contract: every failure mode degrades through the spec §7.1
//! ladder; the engine always returns a usable [`EgressRate`].

use std::collections::HashSet;
use std::str::FromStr;
use std::sync::{LazyLock, Mutex};

use rust_decimal::Decimal;
use serde_json::Value;

/// The bundled catalog, embedded at compile time. Mirrors Python's
/// `importlib.resources` lookup.
const BUNDLED_CATALOG: &str = include_str!("../data/egress_prices.json");

/// Tier-4 ultimate fallback — used only when the catalog cannot be read at
/// all AND `_meta.default_rate_usd_per_gb` cannot be resolved. Matches the
/// spec §7.1 hardcoded constant.
fn hardcoded_default() -> Decimal {
    Decimal::from_str_exact("0.09").expect("0.09 is a valid Decimal")
}

/// Module-level warn-once tracking, keyed by failure-mode token.
static WARNED_MODES: LazyLock<Mutex<HashSet<String>>> =
    LazyLock::new(|| Mutex::new(HashSet::new()));

fn warn_once(mode: &str, message: &str) {
    let mut guard = WARNED_MODES.lock().expect("warn-once mutex poisoned");
    if guard.contains(mode) {
        return;
    }
    guard.insert(mode.to_string());
    drop(guard);
    eprintln!("[dexcost] WARNING: {}", message);
}

/// Test-only: clear the warn-once tracking set.
#[doc(hidden)]
pub fn reset_warning_state_for_tests() {
    let mut guard = WARNED_MODES.lock().expect("warn-once mutex poisoned");
    guard.clear();
}

/// The result of an egress-rate lookup.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EgressRate {
    pub rate_per_gb: Decimal,
    pub pricing_source: String,
    /// One of `"exact"`, `"computed"`, `"estimated"`.
    pub cost_confidence: String,
}

/// Resolve egress rates from the bundled catalog.
#[derive(Debug, Clone)]
pub struct EgressPricingEngine {
    catalog: Value,
    catalog_version: String,
}

impl Default for EgressPricingEngine {
    fn default() -> Self {
        Self::new()
    }
}

impl EgressPricingEngine {
    /// Loads from the bundled `egress_prices.json`.
    pub fn new() -> Self {
        Self::from_str(BUNDLED_CATALOG)
    }

    /// Loads from a caller-supplied JSON string (used for test overrides and
    /// the "catalog unreadable" tiers).
    #[allow(clippy::should_implement_trait)]
    pub fn from_str(catalog_json: &str) -> Self {
        let catalog: Value = match serde_json::from_str(catalog_json) {
            Ok(v) => v,
            Err(e) => {
                warn_once(
                    "catalog_malformed",
                    &format!(
                        "egress catalog malformed JSON ({}); falling back to hardcoded default",
                        e
                    ),
                );
                return Self {
                    catalog: Value::Null,
                    catalog_version: "unknown".to_string(),
                };
            }
        };

        let version = catalog
            .get("_meta")
            .and_then(|m| m.get("version"))
            .and_then(|v| v.as_str())
            .unwrap_or("unknown")
            .to_string();

        Self {
            catalog,
            catalog_version: version,
        }
    }

    /// Loads a catalog from a filesystem path. If the file is missing or
    /// unreadable, the engine falls through to the hardcoded default at lookup
    /// time. Mirrors the `catalog_path` override in the Python engine.
    pub fn from_path(path: &std::path::Path) -> Self {
        match std::fs::read_to_string(path) {
            Ok(raw) => Self::from_str(&raw),
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
                warn_once(
                    "catalog_missing",
                    "egress catalog file not found; falling back to hardcoded default",
                );
                Self {
                    catalog: Value::Null,
                    catalog_version: "unknown".to_string(),
                }
            }
            Err(e) => {
                warn_once(
                    "catalog_unreadable",
                    &format!(
                        "egress catalog unreadable ({}); falling back to hardcoded default",
                        e
                    ),
                );
                Self {
                    catalog: Value::Null,
                    catalog_version: "unknown".to_string(),
                }
            }
        }
    }

    pub fn catalog_version(&self) -> &str {
        &self.catalog_version
    }

    /// Rate for a call classified as internal traffic — always free.
    pub fn rate_for_internal(&self) -> EgressRate {
        EgressRate {
            rate_per_gb: Decimal::ZERO,
            pricing_source: "egress_catalog:internal".to_string(),
            cost_confidence: "exact".to_string(),
        }
    }

    /// Resolve an egress rate via the §7.1 degradation ladder.
    ///
    /// - Tier 1: `(provider, region)` exact catalog match → region rate, `computed`
    /// - Tier 2: provider known, region missing → provider default, `estimated`
    /// - Tier 3: provider unknown / not in catalog → `_meta` default, `estimated`
    /// - Tier 4: catalog unreadable / meta default missing → hardcoded `0.09`, `estimated`
    pub fn resolve_rate(&self, provider: Option<&str>, region: Option<&str>) -> EgressRate {
        if let Some(prov) = provider {
            if let Some(block) = self.catalog.get(prov) {
                if block.is_object() {
                    // Tier 1: exact (provider, region) match.
                    if let Some(reg) = region {
                        if let Some(rate_val) =
                            block.get("regions").and_then(|r| r.get(reg))
                        {
                            let s = match rate_val {
                                Value::String(s) => Some(s.as_str().to_string()),
                                Value::Number(n) => Some(n.to_string()),
                                _ => None,
                            };
                            if let Some(s) = s {
                                match Decimal::from_str(&s) {
                                    Ok(rate) => {
                                        return EgressRate {
                                            rate_per_gb: rate,
                                            pricing_source: format!(
                                                "egress_catalog:{}:{}",
                                                prov, reg
                                            ),
                                            cost_confidence: "computed".to_string(),
                                        };
                                    }
                                    Err(_) => {
                                        warn_once(
                                            &format!(
                                                "region_rate_malformed:{}:{}",
                                                prov, reg
                                            ),
                                            &format!(
                                                "egress region rate malformed for {}/{}",
                                                prov, reg
                                            ),
                                        );
                                    }
                                }
                            }
                        }
                    }

                    // Tier 2: provider default.
                    if let Some(d) = block.get("default_usd_per_gb") {
                        let s = match d {
                            Value::String(s) => Some(s.as_str().to_string()),
                            Value::Number(n) => Some(n.to_string()),
                            _ => None,
                        };
                        if let Some(s) = s {
                            if let Ok(rate) = Decimal::from_str(&s) {
                                return EgressRate {
                                    rate_per_gb: rate,
                                    pricing_source: format!(
                                        "egress_catalog:{}:default",
                                        prov
                                    ),
                                    cost_confidence: "estimated".to_string(),
                                };
                            }
                        }
                    }
                }
            }
        }

        // Tier 3: _meta default.
        if let Some(meta) = self.catalog.get("_meta") {
            if let Some(d) = meta.get("default_rate_usd_per_gb") {
                let s = match d {
                    Value::String(s) => Some(s.as_str().to_string()),
                    Value::Number(n) => Some(n.to_string()),
                    _ => None,
                };
                if let Some(s) = s {
                    if let Ok(rate) = Decimal::from_str(&s) {
                        return EgressRate {
                            rate_per_gb: rate,
                            pricing_source: "egress_catalog:default".to_string(),
                            cost_confidence: "estimated".to_string(),
                        };
                    }
                }
            }
            warn_once(
                "meta_default_missing",
                "egress _meta.default_rate_usd_per_gb missing/malformed; \
                 using hardcoded default",
            );
        }

        // Tier 4: hardcoded.
        EgressRate {
            rate_per_gb: hardcoded_default(),
            pricing_source: "egress_catalog:default".to_string(),
            cost_confidence: "estimated".to_string(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::Mutex as StdMutex;

    /// Tests touch process-global warn-once state; serialize them so a
    /// `reset_warning_state_for_tests` call in one test doesn't race
    /// another's assertion.
    static TEST_LOCK: LazyLock<StdMutex<()>> = LazyLock::new(|| StdMutex::new(()));

    fn lock() -> std::sync::MutexGuard<'static, ()> {
        match TEST_LOCK.lock() {
            Ok(g) => g,
            Err(poisoned) => poisoned.into_inner(),
        }
    }

    #[test]
    fn test_tier1_region_match_is_computed() {
        let _g = lock();
        reset_warning_state_for_tests();
        let eng = EgressPricingEngine::new();
        let r = eng.resolve_rate(Some("aws"), Some("us-east-1"));
        assert_eq!(r.rate_per_gb, Decimal::from_str("0.09").unwrap());
        assert_eq!(r.pricing_source, "egress_catalog:aws:us-east-1");
        assert_eq!(r.cost_confidence, "computed");
    }

    #[test]
    fn test_tier2_provider_known_region_missing_is_estimated() {
        let _g = lock();
        reset_warning_state_for_tests();
        let eng = EgressPricingEngine::new();
        let r = eng.resolve_rate(Some("aws"), Some("moon-base-1"));
        assert_eq!(r.rate_per_gb, Decimal::from_str("0.09").unwrap());
        assert_eq!(r.pricing_source, "egress_catalog:aws:default");
        assert_eq!(r.cost_confidence, "estimated");
    }

    #[test]
    fn test_tier3_unknown_provider_falls_to_meta_default() {
        let _g = lock();
        reset_warning_state_for_tests();
        let eng = EgressPricingEngine::new();
        let r = eng.resolve_rate(None, None);
        assert_eq!(r.rate_per_gb, Decimal::from_str("0.09").unwrap());
        assert_eq!(r.pricing_source, "egress_catalog:default");
        assert_eq!(r.cost_confidence, "estimated");
    }

    #[test]
    fn test_internal_traffic_is_free_and_exact() {
        let _g = lock();
        reset_warning_state_for_tests();
        let eng = EgressPricingEngine::new();
        let r = eng.rate_for_internal();
        assert_eq!(r.rate_per_gb, Decimal::ZERO);
        assert_eq!(r.pricing_source, "egress_catalog:internal");
        assert_eq!(r.cost_confidence, "exact");
    }

    #[test]
    fn test_tier4_missing_catalog_falls_to_hardcoded() {
        let _g = lock();
        reset_warning_state_for_tests();
        let tmp = tempfile::tempdir().unwrap();
        let bogus = tmp.path().join("no.json");
        let eng = EgressPricingEngine::from_path(&bogus);
        let r = eng.resolve_rate(Some("aws"), Some("us-east-1"));
        assert_eq!(r.rate_per_gb, Decimal::from_str("0.09").unwrap());
        assert_eq!(r.cost_confidence, "estimated");
    }

    #[test]
    fn test_tier4_malformed_catalog_falls_to_hardcoded() {
        let _g = lock();
        reset_warning_state_for_tests();
        let eng = EgressPricingEngine::from_str("{not json");
        let r = eng.resolve_rate(Some("aws"), Some("us-east-1"));
        assert_eq!(r.rate_per_gb, Decimal::from_str("0.09").unwrap());
        assert_eq!(r.cost_confidence, "estimated");
    }

    #[test]
    fn test_tier4_meta_default_missing_falls_to_hardcoded() {
        let _g = lock();
        reset_warning_state_for_tests();
        let json = r#"{"_meta": {"version": "x", "currency": "USD"}}"#;
        let eng = EgressPricingEngine::from_str(json);
        let r = eng.resolve_rate(None, None);
        assert_eq!(r.rate_per_gb, Decimal::from_str("0.09").unwrap());
        assert_eq!(r.cost_confidence, "estimated");
    }

    #[test]
    fn test_warn_once_per_failure_mode() {
        let _g = lock();
        reset_warning_state_for_tests();
        // Build two engines with the SAME failure mode; only one warning
        // should be tracked.
        let _e1 = EgressPricingEngine::from_str("{");
        let _e2 = EgressPricingEngine::from_str("{also bad");
        let guard = WARNED_MODES.lock().unwrap();
        assert_eq!(guard.len(), 1);
        assert!(guard.contains("catalog_malformed"));
    }

    #[test]
    fn test_warn_distinct_modes_independently() {
        let _g = lock();
        reset_warning_state_for_tests();
        let tmp = tempfile::tempdir().unwrap();
        let missing = tmp.path().join("missing.json");
        let _e1 = EgressPricingEngine::from_path(&missing);
        let _e2 = EgressPricingEngine::from_str("{");
        let guard = WARNED_MODES.lock().unwrap();
        assert!(guard.contains("catalog_missing"));
        assert!(guard.contains("catalog_malformed"));
    }

    #[test]
    fn test_decimal_no_float_drift() {
        // Critical invariant: the per-GB divisor and rate math must be exact.
        // 1e9 in f64 has no representable drift, but multiplying user rates
        // by a Decimal divisor is the contract.
        let a = Decimal::from_str("0.1093").unwrap()
            * Decimal::from(1_000_000_000_u64);
        assert_eq!(a, Decimal::from_str("109300000.0000").unwrap());

        let b = Decimal::from_str("0.087").unwrap()
            * Decimal::from(12_345_678_u64);
        assert_eq!(b, Decimal::from_str("1074073.986").unwrap());
    }

    #[test]
    fn test_pricing_version_from_meta() {
        let _g = lock();
        reset_warning_state_for_tests();
        let eng = EgressPricingEngine::new();
        assert_eq!(eng.catalog_version(), "1.0.0");
    }

    #[test]
    fn test_egress_rate_is_clone_eq() {
        // Rust analog of Python's "frozen dataclass" — EgressRate derives
        // Clone/PartialEq/Eq and has no interior mutability.
        let _g = lock();
        reset_warning_state_for_tests();
        let eng = EgressPricingEngine::new();
        let r = eng.resolve_rate(Some("aws"), Some("us-east-1"));
        let r2 = r.clone();
        assert_eq!(r, r2);
    }
}

File: python/src/dexcost/egress_pricing.py
"""Egress pricing engine — resolves a per-GB egress rate from
``(provider, region)`` using the bundled ``data/egress_prices.json`` catalog.

Mirrors :mod:`dexcost.pricing` in shape: bundled JSON + a resolver returning
``(rate, pricing_source, cost_confidence)``.

Fail-silent contract: every failure mode degrades through the spec §7.1
ladder; the engine always returns a usable :class:`EgressRate`.
"""

from __future__ import annotations

import json
import logging
import threading
from dataclasses import dataclass
from decimal import Decimal, InvalidOperation
from importlib import resources
from pathlib import Path
from typing import Any

_log = logging.getLogger(__name__)

# Tier-4 ultimate fallback — used only when the catalog cannot be read at all
# AND _meta.default_rate_usd_per_gb cannot be resolved.  Matches the spec §7.1
# hardcoded constant.
_HARDCODED_DEFAULT = Decimal("0.09")

_warned_modes: set[str] = set()
_warn_lock = threading.Lock()


def _reset_warning_state() -> None:
    """Test-only: clear the warn-once tracking set."""
    with _warn_lock:
        _warned_modes.clear()


def _warn_once(mode: str, message: str) -> None:
    with _warn_lock:
        if mode in _warned_modes:
            return
        _warned_modes.add(mode)
    _log.warning(message)


@dataclass(frozen=True)
class EgressRate:
    """The result of an egress-rate lookup."""

    rate_per_gb: Decimal
    billing_unit: str
    pricing_source: str
    cost_confidence: str  # exact | computed | estimated


class EgressPricingEngine:
    """Resolve egress rates from the bundled catalog.

    Args:
        catalog_path: Optional override path. ``None`` uses the bundled
            ``data/egress_prices.json``.
    """

    def __init__(
        self,
        catalog_path: str | Path | None = None,
        *,
        catalog_data: dict[str, Any] | None = None,
        catalog_version: str | None = None,
        rate_overrides: dict[str, tuple[Decimal, str]] | None = None,
    ) -> None:
        if catalog_path is not None and catalog_data is not None:
            raise ValueError("catalog_path and catalog_data are mutually exclusive")
        self._catalog: dict[str, Any] = {}
        self._catalog_path = catalog_path
        self._catalog_data = catalog_data
        self._catalog_version_override = catalog_version
        self._rate_overrides = rate_overrides or {}
        self._catalog_version: str = "unknown"
        self._load()

    def _load(self) -> None:
        try:
            if self._catalog_data is not None:
                raw = json.dumps(self._catalog_data, sort_keys=True, separators=(",", ":"))
            elif self._catalog_path is not None:
                raw = Path(self._catalog_path).read_text(encoding="utf-8")
            else:
                raw = (
                    resources.files("dexcost")
                    .joinpath("data")
                    .joinpath("egress_prices.json")
                    .read_text(encoding="utf-8")
                )
        except FileNotFoundError:
            _warn_once(
                "catalog_missing",
                "egress catalog file not found; falling back to hardcoded default",
            )
            return
        except OSError as exc:
            _warn_once(
                "catalog_unreadable",
                f"egress catalog unreadable ({exc}); falling back to hardcoded default",
            )
            return

        try:
            self._catalog = json.loads(raw)
        except json.JSONDecodeError as exc:
            _warn_once(
                "catalog_malformed",
                f"egress catalog malformed JSON ({exc}); falling back to hardcoded default",
            )
            self._catalog = {}
            return

        meta = self._catalog.get("_meta", {})
        self._catalog_version = self._catalog_version_override or str(
            meta.get("version", "unknown")
        )

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    @property
    def catalog_version(self) -> str:
        return self._catalog_version

    def rate_for_internal(self) -> EgressRate:
        """Rate for a call classified as internal traffic — always free."""
        return EgressRate(
            rate_per_gb=Decimal("0"),
            billing_unit="gb_egress",
            pricing_source="egress_catalog:internal",
            cost_confidence="exact",
        )

    def resolve_rate(self, provider: str | None, region: str | None) -> EgressRate:
        """Resolve an egress rate via the §7.1 degradation ladder.

        Tier 1: ``(provider, region)`` exact match → region rate, ``computed``.
        Tier 2: provider known, region absent/unknown → provider default,
            ``estimated``.
        Tier 3: provider not detected / not in catalog → ``_meta`` default,
            ``estimated``.
        Tier 4: catalog unreadable or ``_meta`` default absent → hardcoded
            ``Decimal("0.09")``, ``estimated``.
        """
        if provider:
            keys = [f"{provider}:{region}", provider] if region else [provider]
            for key in keys:
                override = self._rate_overrides.get(key)
                if override is not None:
                    amount, billing_unit = override
                    return EgressRate(
                        rate_per_gb=amount,
                        billing_unit=billing_unit,
                        pricing_source=(
                            f"workspace_overlay:egress:{key}:{billing_unit}"
                        ),
                        cost_confidence="computed",
                    )
        if provider:
            block = self._catalog.get(provider)
            if isinstance(block, dict):
                regions = block.get("regions", {})
                if region and region in regions:
                    try:
                        rate = Decimal(str(regions[region]))
                    except InvalidOperation:
                        _warn_once(
                            f"region_rate_malformed:{provider}:{region}",
                            f"egress region rate malformed for {provider}/{region}",
                        )
                    else:
                        return EgressRate(
                            rate_per_gb=rate,
                            billing_unit="gb_egress",
                            pricing_source=f"egress_catalog:{provider}:{region}",
                            cost_confidence="computed",
                        )
                prov_default: Decimal | None
                try:
                    prov_default = Decimal(str(block.get("default_usd_per_gb", "")))
                except (InvalidOperation, TypeError):
                    prov_default = None
                if prov_default is not None:
                    return EgressRate(
                        rate_per_gb=prov_default,
                        billing_unit="gb_egress",
                        pricing_source=f"egress_catalog:{provider}:default",
                        cost_confidence="estimated",
                    )

        meta = self._catalog.get("_meta") if self._catalog else None
        if isinstance(meta, dict):
            meta_rate: Decimal | None
            try:
                meta_rate = Decimal(str(meta.get("default_rate_usd_per_gb", "")))
            except (InvalidOperation, TypeError):
                meta_rate = None
            if meta_rate is not None:
                return EgressRate(
                    rate_per_gb=meta_rate,
                    billing_unit="gb_egress",
                    pricing_source="egress_catalog:default",
                    cost_confidence="estimated",
                )
            _warn_once(
                "meta_default_missing",
                "egress _meta.default_rate_usd_per_gb missing/malformed; "
                "using hardcoded default",
            )

        return EgressRate(
            rate_per_gb=_HARDCODED_DEFAULT,
            billing_unit="gb_egress",
            pricing_source="egress_catalog:default",
            cost_confidence="estimated",
        )
